src/core/params.hpp

```diff
@@ -0,0 +1,138 @@
+
+  template <typename T> static IterableParam create(size_t offset, const ParameterInfo *paramInfo) {
+    return createWithVarInterface<T>(offset, paramInfo);
+  }
+
+  template <typename T> static IterableParam createWithVarInterface(size_t offset, const ParameterInfo *paramInfo) {
+    IterableParam result{
+        .offset = offset,
+        .paramInfo = paramInfo,
+        .setParam = [](void *varPtr, SHVar var) { *((T *)varPtr) = var; },
+        .getParam = [](void *varPtr) -> SHVar { return *((T *)varPtr); },
+    };
+
+    if constexpr (has_warmup<T>::value) {
+      result.warmup = [](void *varPtr, SHContext *ctx) { ((T *)varPtr)->warmup(ctx); };
+    } else {
+      result.warmup = [](void *, SHContext *) {};
+    }
+
+    if constexpr (has_warmup<T>::value) {
+      result.cleanup = [](void *varPtr) { ((T *)varPtr)->cleanup(); };
+    } else {
+      result.cleanup = [](void *) {};
+    }
+
+    return result;
+  }
+};
+
+// Usage:
+//  PARAM_IMPL(Shard,
+//    PARAM_IMPL_FOR(Param0),
+//    PARAM_IMPL_FOR(Param1),
+//    PARAM_IMPL_FOR(Param2))
+// Side effects:
+//  - typedefs Self to the current class
+//  - creates a static member named iterableParams
+#define PARAM_IMPL(_type, ...)                                          \
+  typedef _type Self;                                                   \
+  static const IterableParam *getIterableParams(size_t &outNumParams) { \
+    static IterableParam result[] = {__VA_ARGS__};                      \
+    outNumParams = std::extent<decltype(result)>::value;                \
+    return result;                                                      \
+  }                                                                     \
+  PARAM_PARAMS()                                                        \
+  PARAM_GET_SET()
+
+#define PARAM_IMPL_FOR(_name) IterableParam::create<decltype(_name)>(offsetof(Self, _name), &_name##ParameterInfo)
+
+// Implements parameters()
+#define PARAM_PARAMS()                                                          \
+  static SHParametersInfo parameters() {                                        \
+    static SHParametersInfo result = []() {                                     \
+      SHParametersInfo result{};                                                \
+      size_t numParams;                                                         \
+      const IterableParam *params = getIterableParams(numParams);               \
+      arrayResize(result, numParams);                                           \
+      for (size_t i = 0; i < numParams; i++) {                                  \
+        result.elements[i] = *const_cast<ParameterInfo *>(params[i].paramInfo); \
+      }                                                                         \
+      return result;                                                            \
+    }();                                                                        \
```

Comment:
C++ question: does using a closure like this still make it statically allocated (i.e. resolved at compile time) or dynamically allocated? In the later case, does it mean that the resulting binary is potentially smaller since memory will only be determined at runtime?

Reply:
There are no allocations at all.
Compiler will create a struct with the captured things, copy them inside and feed them to a call.

Reply:
Moreover nothing is captured here

Reply:
Oh, I think you meant using the call like that. This initializes it at runtime, but still only once (lazy initialization)

Reply:
Yes my question was regarding `static SHParametersInfo result` allocation.